hwmv2: Fix problems with common board functionality

nordicjm opened the pull request:
This change reverts a change that was introduced with hwmv2 which
    allowed for using common board names for overlays, given the board
    target ``nrf9160dk@0.7.0/nrf9160/ns``.
    
    In hwmv1 this would have used:
        ``nrf9160dk_nrf9160_ns.conf``.
        ``nrf9160dk_nrf9160_ns_0_7_0.conf``.
    
    In hwmv2 this would have used:
        ``nrf9160dk_nrf9160_ns.conf``
        ``nrf9160dk_nrf9160_ns_0_7_0.conf``
        ``nrf9160dk_nrf9160.conf``
        ``nrf9160dk_nrf9160_0_7_0.conf``
        ``nrf9160dk.conf``
        ``nrf9160dk_0_7_0.conf``
    
    With these changes, the following are used (which restores the hwmv1
    behaviour):
        ``nrf9160dk_nrf9160_ns.conf``
        ``nrf9160dk_nrf9160_ns_0_7_0.conf``
    
    For a board with a default SoC which is not a variant for example
    ``rpi_pico`` then ``rpi_pico.conf`` will also be used, this file
    will not be used for variants e.g. ``rpi_pico/rp2040/w``
    
    This applies to .dts, .conf and .overlay files in the boards
    directory, and to .conf and .overlay files in application board
    overlay directories.
    
    This revert is needed to avoid issues whereby variants have
    incompatible configuration to the parent board target, which has
    been affecting samples and tests.


    cmake: Deprecate board files that do not have SoC in the name
    
    Deprecated .dts, .overlay and .conf files that have the name of the
    board itself but without the name of the SoC in them, emits cmake
    deprecation notices upon such files being found.


**Future plans for missing functionality (provisional)**

We have internally discussed the following which could be added in future to enhance the features removed/changed here:

- Having twister support short board names as cmake/`west build` does, by having it use the build system to determine if a name is valid
- Having common file support but marking the files explicitly as common files, the current idea is to use a suffix in the filename so `nrf5340dk_nrf5340_cpuapp.conf` will only apply to `nrf5340dk/nrf5340/cpuapp` but something like `nrf5340dk_nrf5340_cpuapp.all.conf` could apply to this board target and all variants: `nrf5340dk/nrf5340/cpuapp` and `nrf5340dk/nrf5340/cpuapp/ns`

Fixes #70445
Fixed #70283

nordicjm committed:
cmake: modules: extensions: Revert using common board files

This change reverts a change that was introduced with hwmv2 which
allowed for using common board names for overlays, given the board
target ``nrf9160dk@0.7.0/nrf9160/ns``.

In hwmv1 this would have used:
    ``nrf9160dk_nrf9160_ns.conf``.
    ``nrf9160dk_nrf9160_ns_0_7_0.conf``.

In hwmv2 this would have used:
    ``nrf9160dk_nrf9160_ns.conf``
    ``nrf9160dk_nrf9160_ns_0_7_0.conf``
    ``nrf9160dk_nrf9160.conf``
    ``nrf9160dk_nrf9160_0_7_0.conf``
    ``nrf9160dk.conf``
    ``nrf9160dk_0_7_0.conf``

With these changes, the following are used (which restores the hwmv1
behaviour):
    ``nrf9160dk_nrf9160_ns.conf``
    ``nrf9160dk_nrf9160_ns_0_7_0.conf``

For a board with a default SoC which is not a variant for example
``rpi_pico`` then ``rpi_pico.conf`` will also be used, this file
will not be used for variants e.g. ``rpi_pico/rp2040/w``

This applies to .dts, .conf and .overlay files in the boards
directory, and to .conf and .overlay files in application board
overlay directories.

This revert is needed to avoid issues whereby variants have
incompatible configuration to the parent board target, which has
been affecting samples and tests.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
Revert "boards: Raspberry Pi pico pwm led adjustment"

This reverts commit 037a3b52a49a8e26ac3ee4bdac46c72d76bc7399.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
Revert "samples: blinky_pwm: enable pwm_leds in rpi_pico overlay"

This reverts commit 7277cae6fa7c3d0ad567b4f1ef9007dc12e44314.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1593,20 +1593,17 @@ function(zephyr_build_string outvar)
-      list(POP_BACK str_segment_list)
-    endwhile()
+
+    list(LENGTH str_segment_list str_segment_list_len)
+    if("${str_segment_list_len}" EQUAL "2")
+      # Include board name without SoC
+      list(APPEND ${outvar} "${BUILD_STR_BOARD}")
+    endif()
```
initially I'm not sure we should change the behavior of `MERGE`.
Instead we should probably avoid using `MERGE` in this case, and instead update code here:
https://github.com/zephyrproject-rtos/zephyr/blob/62665eabf7d29bbdf3309e197d3d01d67b669b21/cmake/modules/configuration_files.cmake#L79-L89

That code can then do something like:
```
  zephyr_build_string(board_overlay_string
                      BOARD ${BOARD}
                      BOARD_IDENTIFIER ${BOARD_IDENTIFIER}
  )
  set(board_overlay_string "${board_overlay_string}.overlay")
  if(NOT EXISTS "${board_overlay_string}")
    string(REGEX REPLACE" "^[^/]*/" "" legacy_board_identifier "${BOARD_IDENTIFIER}"
​    zephyr_build_string(legacy_board_overlay_string
                        BOARD ${BOARD}
                        BOARD_IDENTIFIER ${legacy_board_identifier}
    )
    set(legacy_board_overlay_string "${legacy_board_overlay_string}.overlay")
    if(EXISTS "${legacy_board_overlay_string}.overlay")
        message(WARNING Overlay file using legacy board target file name structure (${legacy_board_overlay_string}) detected.
         Rename overlay to: ${board_overlay_string})
        set(board_overlay_string ${legacy_board_overlay_string})
    endif()
  endif()
  
  zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR} DTS DTC_OVERLAY_FILE
              NAMES "${APP_BOARD_DTS};${board_overlay_strings};app.overlay" SUFFIX ${FILE_SUFFIX})
```

and do similar for conf files.

After two releases, then warning code can be removed.
(We can already place the warning code in deprecated.cmake now, but that's outside the point of the example given)

nordicjm replied:
The changes here also apply to board files (i.e. in `boards`), I don't remember if it was raised an issue on github or discord or internally but an issue was raised about common files setting things that then conflicted with configuration for variants, one was for trustzone, another was for qemu, the changes here are intentionally applied to the base function.

tejlmand replied:
> The changes here also apply to board files (i.e. in boards)

Yes, and that would need to be handled in the lines just above:
https://github.com/zephyrproject-rtos/zephyr/blob/62665eabf7d29bbdf3309e197d3d01d67b669b21/cmake/modules/configuration_files.cmake#L76

and `zephyr_file()` should no longer apply `MERGE`, but do identical to Zephyr v3.6.0, which was:
https://github.com/zephyrproject-rtos/zephyr/blob/468eb56cf242eedba62006ee758700ee6148763f/cmake/modules/extensions.cmake#L2550-L2554

instead of current main:
https://github.com/zephyrproject-rtos/zephyr/blob/2396b01b1ea2b0b6a61591155508cf19cd9d367c/cmake/modules/extensions.cmake#L2607

Of course the change to use BOARD_QUALIFIERS after v3.6.0 should be kept.
Only MERGE to be removed.

nordicjm replied:
I don't think this should be done, taking the rpi_pico board, that is the board name, there is no other soc for this device and the twister identifier for this in hwmv2 is as it was in hwmv2: `rpi_pico`, the `w` variant had to have the name updated to `rpi_pico/rp2040/w` because that is required for variants but this would be a drastic change which then would invalidate (deprecate) **all** board names for in-tree boards (which are also used throughout tests and samples) which do not have the soc in the name. I would like @nashif to give feedback on if this is acceptable

tejlmand replied:
> but this would be a drastic change which then would invalidate (deprecate) all board names for in-tree boards (which are also used throughout tests and samples) which do not have the soc in the name. 

We need to make it clear how board overlays are being picked up.
Either always merge, as we do now, which clearly have drawbacks. 
Or only used the normalized board target name, where we don't merge at all, but allow legacy format for a given time (that is file names without the SoC qualifier).

A combination where you would just allow SoC to be omitted but require the CPU set and variant parts will create a mess and also have edge-cases to deal with in future. (should a file be name `plank__for.overlay` (equivalent to `BOARD=plank//foo`) or just `plank_foo.overlay` ?

The possibility to omit SoC on command line invocation is a user convenience, but there is no reason why a fixed file name should be allowed to do so, because naming a file is a one time operation. 

We seem to have accepted that the ability to inherit overlay files was a feature with undesired side-effects, and therefore long term there is no reason to allow different ways to name the same file.

Of course we don't need to print the warning immediately, but can allow a longer grace period for transitioning the overlay file names.

nordicjm replied:
So before I start working on this I want agreement from other that this is going to be accepted, the following are all board identifiers which do not have the soc in the name, so all these board identifiers will need to change, and any overlay or Kconfig fragments for these will need to be renamed:
```
96b_aerocore2
96b_argonkey
96b_avenger96
96b_neonkey
96b_nitrogen
96b_stm32_sensor_mez
96b_wistrio
acn52832
acrn
acrn_ehl_crb
adafruit_feather_m0_basic_proto
adafruit_feather_m0_lora
adafruit_feather_stm32f405
adafruit_grand_central_m4_express
adafruit_itsybitsy_m4_express
adafruit_kb2040
nrf52_adafruit_feather
adafruit_qt_py_rp2040
adafruit_trinket_m0
adi_eval_adin1110ebz
adi_eval_adin2111ebz
adi_sdp_k1
pandora_stm32l475
altera_max10
apollo4p_blue_kxr_evb
apollo4p_evb
kv260_r5
arduino_due
arduino_mkrzero
arduino_nano_33_ble
arduino_nano_33_iot
arduino_nicla_sense_me
arduino_uno_r4_minima
arduino_zero
fvp_base_revc_2xaemv8a
v2m_beetle
ast1030_evb
sam4e_xpro
sam4l_ek
sam4s_xplained
samc21n_xpro
samd20_xpro
samd21_xpro
same54_xpro
saml21_xpro
samr21_xpro
samr34_xpro
bbc_microbit
bbc_microbit_v2
beagleconnect_freedom
beaglev_fire
swan_r5
bcm958401m2
xt-sim
cy8ckit_062s4
cy8cproto_062_4343w
cy8cproto_063_ble
zybo
dragino_lsn50
dragino_nbsn95
titanium_ti60_f225
mercury_xu
kb1200_evb
esp32c3_devkitm
esp32s2_saola
fk7b0m1_vbt6
roc_rk3568_pc
esp32s2_franzininho
generic_leon3
gr716a_mini
gd32a503v_eval
gd32e103v_eval
gd32e507v_start
gd32e507z_eval
gd32f350r_eval
gd32f403z_eval
gd32f407v_start
gd32f450i_eval
gd32f450v_start
gd32f450z_eval
gd32f470i_eval
gd32l233r_eval
gd32vf103c_starter
gd32vf103v_eval
google_dragonclaw
google_kukui
google_twinkie_v2
holyiot_yj16019
xmc45_relax_kit
xmc47_relax_kit
intel_adl_crb
intel_adl_rvp
intel_ehl_crb
intel_ish_5_4_1
intel_ish_5_6_0
intel_ish_5_8_0
niosv_g
niosv_m
intel_rpl_p_crb
intel_rpl_s_crb
intel_socfpga_agilex5_socdk
intel_socfpga_agilex_socdk
cyclonev_socdk
it82xx2_evb
it8xxx2_evb
khadas_edgev
bl652_dvk
bl653_dvk
bl654_dvk
bl654_sensor_board
bl654_usb
bt510
bt610
mg100
pinnacle_100_dvk
rm1xx_dvk
opentitan_earlgrey
stamp_c3
mm_feather
mm_swiftio
ev11l78a
m2gl025_miv
mec1501modular_assy6885
mec15xxevb_assy6853
mec172xevb_assy6906
mec172xmodular_assy6930
mpfs_icicle
mikroe_clicker_2
mikroe_mini_m4_for_stm32
az3166_iotdevkit
native_posix
native_sim
nrf52_bsim
nrf5340bsim_nrf5340_cpuapp
nrf5340bsim_nrf5340_cpunet
npcx4m8f_evb
npcx7m6fb_evb
npcx9m6f_evb
numaker_pfm_m467
numaker_pfm_m487
numaker_m2l31ki
frdm_k22f
frdm_k64f
frdm_k82f
frdm_kl25z
frdm_kw41z
lpcxpresso11u68
lpcxpresso51u68
lpcxpresso55s06
lpcxpresso55s16
lpcxpresso55s28
lpcxpresso55s36
ls1046ardb
mimxrt1010_evk
mimxrt1015_evk
mimxrt1020_evk
mimxrt1024_evk
mimxrt1040_evk
mimxrt1050_evk
mimxrt1050_evk@qspi
mimxrt1060_evk@hyperflash
mimxrt1060_evk
mimxrt1060_evkb
mimxrt1062_fmurt6
mimxrt1064_evk
mimxrt685_evk
mr_canhubk3
rd_rw612_bga
rddrone_fmuk66
twr_ke18f
twr_kv58f220m
ucans32k1sic
usb_kw24d512
olimex_lora_stm32wl_devkit
olimex_lora_stm32wl_devkit@D
olimexino_stm32
olimex_stm32_e407
olimex_stm32_h103
olimex_stm32_h405
olimex_stm32_h407
olimex_stm32_p405
black_f407ve
black_f407zg_pro
icev_wireless
neorv32
stm32_min_dev@black
stm32_min_dev@blue
stm32f030_demo
stm32f401_mini
serpente
pan1770_evb
pan1780_evb
pan1781_evb
pan1782_evb
particle_argon
particle_boron
nrf51_blenano
nrf52_blenano2
particle_xenon
reel_board
reel_board@2
pinetime_devkit0
teensy40
teensy41
qemu_cortex_a53
qemu_cortex_a9
qemu_cortex_m0
qemu_cortex_m3
qemu_cortex_r5
qemu_kvm_arm64
qemu_leon3
qemu_malta
qemu_nios2
qemu_riscv32
qemu_riscv32_xip
qemu_riscv32e
qemu_riscv64
qemu_x86
qemu_x86_64
qemu_x86_lakemont
qemu_x86_tiny
qemu_xtensa
decawave_dwm1001_dev
qomu
quick_feather
rpi_4b
rpi_pico
da14695_dk_usb
da1469x_dk_pro
rcar_salvator_xs
rcar_spider_s4
rzt2m_starter_kit
riscv32_virtual
ronoth_lodev
ruuvi_ruuvitag
scobc_module1
faze
legend@25hdd
legend@25ssd
legend@35
stm32f3_seco_d23
lora_e5_dev_board
lora_e5_mini
seeeduino_xiao
wio_terminal
xiao_ble
xiao_esp32c3
ip_k66f
segger_trb_stm32f407
hifive1
hifive1@B
hifive_unleashed
hifive_unmatched
efm32gg_sltb009a
efm32gg_slwstk6121a
efm32gg_stk3701a
efm32hg_slstk3400a
efm32pg_stk3401a
efm32wg_stk3800
efr32bg22_brd4184a
efr32bg22_brd4184b
efr32bg27_brd2602a
efr32mg_sltb004a
efr32xg24_dk2601b
longan_nano
nrf52_sparkfun
sparkfun_pro_micro_rp2040
sparkfun_red_v_things_plus
b_g474e_dpow1
b_l072z_lrwan1
b_l4s5i_iot01a
b_u585i_iot02a
disco_l475_iot1
nucleo_c031c6
nucleo_f030r8
nucleo_f030r8@2
nucleo_f031k6
nucleo_f042k6
nucleo_f070rb
nucleo_f091rc
nucleo_f103rb
nucleo_f207zg
nucleo_f302r8
nucleo_f303k8
nucleo_f303re
nucleo_f334r8
nucleo_f401re
nucleo_f410rb
nucleo_f411re
nucleo_f412zg
nucleo_f413zh
nucleo_f429zi
nucleo_f446re
nucleo_f446ze
nucleo_f722ze
nucleo_f746zg
nucleo_f756zg
nucleo_f767zi
nucleo_g031k8
nucleo_g070rb
nucleo_g071rb
nucleo_g0b1re
nucleo_g431rb
nucleo_g474re
nucleo_h563zi
nucleo_h723zg
nucleo_h743zi
nucleo_h753zi
nucleo_h7a3zi_q
nucleo_l011k4
nucleo_l031k6
nucleo_l053r8
nucleo_l073rz
nucleo_l152re
nucleo_l412rb_p
nucleo_l432kc
nucleo_l433rc_p
nucleo_l452re
nucleo_l476rg
nucleo_l496zg
nucleo_l4a6zg
nucleo_l4r5zi
nucleo_l552ze_q
nucleo_u575zi_q
nucleo_u5a5zj_q
nucleo_wb55rg
nucleo_wba52cg
nucleo_wba55cg
nucleo_wl55jc
sensortile_box
sensortile_box_pro
st25dv_mb1283_disco
steval_fcu001v1
stm3210c_eval
stm32373c_eval
stm32f072_eval
stm32f072b_disco
stm32f0_disco
stm32f103_mini
stm32f3_disco
stm32f3_disco@E
stm32f411e_disco@B
stm32f411e_disco
stm32f412g_disco
stm32f429i_disc1
stm32f469i_disco
stm32f4_disco
stm32f723e_disco
stm32f746g_disco
stm32f7508_dk
stm32f769i_disco
stm32g0316_disco
stm32g071b_disco
stm32g081b_eval
stm32h573i_dk
stm32h735g_disco
stm32h750b_dk
stm32h7b3i_dk
stm32l1_disco
stm32l476g_disco
stm32l496g_disco
stm32l4r9i_disco
stm32l562e_dk
stm32mp157c_dk2
stm32u5a9j_dk
stm32vl_disco
stm32wb5mm_dk
stm32wb5mmg
visionfive2
robokit1
tlsr9518adk80d
cc1352p1_launchxl
cc1352r1_launchxl
cc1352r_sensortag
cc26x2r1_launchxl
cc3220sf_launchxl
cc3235sf_launchxl
msp_exp432p401r_launchxl
nrf51_ble400
waveshare_open103z
blackpill_f401cc
blackpill_f401ce
blackpill_f411ce
weact_stm32g431_core
esp32s2_lolin_mini
w5500_evb_pico
xenvm
degu_evk
contextualelectronics_abc
nrf52840_blip
nrf52840_papyr
litex_vexriscv
nrf52832_mdk
nrf52840_mdk
nrf52840_mdk_usb_dongle
hsdk
hsdk4xd
iotdk
up_squared
up_squared_pro_7000
nrf51_vbluno51
nrf52_vbluno52
```

nordicjm replied:
@nandojve @jhedberg @golowanow @rerickson1 @dleach02 @dbaluta @hakehuang @erwango for comment on this, might impact on-device testing

erwango replied:
Why do you quote "only" these targets ? is it because they are today used in samples.yaml and tests.yaml?

If yes, does it mean that basically, you impose the use of `<board>/<soc>` syntax ?

nordicjm replied:
> Why do you quote "only" these targets ? is it because they are today used in samples.yaml and tests.yaml?
> 
> If yes, does it mean that basically, you impose the use of `<board>/<soc>` syntax ?

It would mean having them as the identifier for usage in twister (forced here) and in filenames for overlays e.g. using the example of `nucleo_l031k6` the identifier for twister (and in any testcase.yaml or sample.yaml files) would need to have `nucleo_l031k6/stm32l031xx`, and you might have an application with a `boards/nucleo_l031k6.conf` file, that file name would be deprecated and you would need to use `boards/nucleo_l031k6_stm32l031xx.conf`. The one thing that would not change would be how you use it to build on the command line, so the soc can be omitted as it on supports one soc, thus building using `cmake -GNinja -DBOARD=nucleo_l031k6 <app>` will work

nordicjm replied:
This is not part of the PR (yet) just this discussion thread, want to ensure people are OK with it being added before working on it

erwango replied:
Maybe a usual RFC would be more appropriate to get feedback.
This should sound obvious to you and you probably want to move fast but we're all working on other stuff and we're not up to speed with daily hwmv2 evolutions.

erwango replied:
So, on my side, I'm finding it confusing that we could have targets and twister references that can differ. I expect lot of confusion here, so I'm not in favor of it. What are the options?

nordicjm replied:
5th commit here shows the changes

nandojve replied:
I agreed with @erwango and I'm favor to get the things without any confusing. This means that I would prefer go verbose and have a clear picture about how the overlays etc are selected and how is the order of "merging" content. I don't want to try guess how that could possible work, it should be precise and clear since beginning.

Maybe it could be a little overkill for single core cpu but the complexity grows to much for Multi-core Asymmetric SoC's with TFM for instance. BTW, this new model may allow the development of SOM (Systems On Module) and than we can have multi CPUs etc.

So, if it is necessary to drop complete the hwmv1 compatibility I'm in favor of that too. I think this is a big change and since we are moving to a LTS I prefer not think about hwmv1 anymore, if that is necessary to have a very clear procedure.

tejlmand reviewed (commented):
Perhaps better to issue a warning, while still have temp support for legacy named overlays (and conf files)

nordicjm committed:
boards: Split merged board file configuration up

Splits configuration up that was merged as part of hwmv2 due to
the merged configuration feature being reverted

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

aescolar commented on boards/native/nrf_bsim/nrf52_bsim_native.dts:
If we are to go in this direction in the PR, then I'd ask to do this instead:
```
diff --git a/boards/native/nrf_bsim/board.yml b/boards/native/nrf_bsim/board.yml
index 20c9137a233..5ee927af1e6 100644
--- a/boards/native/nrf_bsim/board.yml
+++ b/boards/native/nrf_bsim/board.yml
@@ -2,7 +2,7 @@ boards:
- name: nrf52_bsim
  vendor: zephyr
  socs:
-  - name: native
+  - name: nrf52833
- name: nrf5340bsim
  vendor: zephyr
  socs:
diff --git a/boards/native/nrf_bsim/nrf52_bsim_native.dts b/boards/native/nrf_bsim/nrf52_bsim_nrf52833.dts
similarity index 100%
rename from boards/native/nrf_bsim/nrf52_bsim_native.dts
rename to boards/native/nrf_bsim/nrf52_bsim_nrf52833.dts
diff --git a/boards/native/nrf_bsim/nrf52_bsim_native.yaml b/boards/native/nrf_bsim/nrf52_bsim_nrf52833.yaml
similarity index 100%
rename from boards/native/nrf_bsim/nrf52_bsim_native.yaml
rename to boards/native/nrf_bsim/nrf52_bsim_nrf52833.yaml
diff --git a/boards/native/nrf_bsim/nrf52_bsim_native_defconfig b/boards/native/nrf_bsim/nrf52_bsim_nrf52833_defconfig
similarity index 100%
rename from boards/native/nrf_bsim/nrf52_bsim_native_defconfig
rename to boards/native/nrf_bsim/nrf52_bsim_nrf52833_defconfig
```
I left the nrf52_bsim with the native soc for not too good reason, and because it was quite hidden that it was that "soc". But if we are to show it all around the place, I'd prefer to have that have the same soc as the real platform.

nordicjm replied:
Updated to `nrf52_bsim/nrf52833`

aescolar commented on boards/native/native_posix/doc/index.rst:
```diff
@@ -11,18 +11,18 @@ Native POSIX execution (native_posix)
 
 But unlike with :ref:`native_sim<native_sim>` you are limited to only using the host C library.
-:ref:`native_sim<native_sim>` supports all ``native_posix`` use cases.
+:ref:`native_sim<native_sim>` supports all ``native_posix/native`` use cases.
```
From an offline discussion with @nordicjm I understand the long term plan is to not allow anymore targeting boards without the soc (in tree, but users may still be able to when building directly(?)). I think this change will be too annoying, as to warrant a discussion in the TSC.

aescolar commented on .github/workflows/codecov.yaml:
```diff
@@ -19,14 +19,14 @@ jobs:
       fail-fast: false
       matrix:
-        platform: ["mps2/an385", "native_sim", "qemu_x86", "unit_testing"]
+        platform: ["mps2/an385", "native_sim/native", "qemu_x86/atom", "unit_testing"]
```
specially as with this change twister would not anymore allow set as target the boards without the SoC name.

nordicjm committed:
manifest: Update MCUboot

Updates MCUboot board names to account for hwmv2 changes

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
cmake: Deprecate board files that do not have SoC in the name

Deprecated .dts, .overlay and .conf files that have the name of the
board itself but without the name of the SoC in them, emits cmake
deprecation notices upon such files being found.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand reviewed (changes requested):
This PR title states: `hwmv2: Revert common board functionality`

and 
> This change reverts a change that was introduced with hwmv2 which
allowed for using common board names for overlays

However, this PR contains a huge amount of changes not related to the title, PR description, or issue #70445.

Please update title and description to properly reflect the changes in this PR.

Besides the linked issue, then that issue purely related to common board in terms of **merging** files without a possibility to opt-out of merging a given file.

There are not issues with sharing of files in the board directory if they are common, like it is for many board dts files, as it is described here:
> Only one .dts file will be used, and the most specific file which exists will be used.
>
> This means that if both plank.dts and plank_soc1_foo.dts exist, then when building for plank / plank/soc1, then plank.dts is used. When building for plank//foo / plank/soc1/foo the plank_soc1_foo.dts is used.

Ref: https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#write-your-devicetree

So please keep renaming of dts files in board dirs out of this PR.

fabiobaltieri commented:
Arch WG:
- @nordicjm presented the issue, currently there are broken configurations in the tree due to how the current implementation picks up configs and merges them
- main concern is that in the new system some configs may be silentely discarded
- twister board names are changing as well, which will impact downstream users as well

nordicjm commented on boards/96boards/nitrogen/96b_nitrogen_nrf52832.dts:
This PR adds 2 files for variants which just include the base, 2 really doesn't seem like an issue to me for the huge benefit of easily being able to see when a .dts file is added which is not valid rather than completing hiding an issue (which in intel's case has not been noticed for an entire month and a half until this PR) and pretending it doesn't exist. So I've checked what files in tree have duplicate contents for a comparison:
```
/tmp/aa/zephyr/boards/native/native_posix/native_posix_native_64.dts - will vanish when old hwmv1 board PR is merged
/tmp/aa/zephyr/boards/native/native_sim/native_sim_native_64.dts  - will vanish when old hwmv1 board PR is merged
/tmp/aa/zephyr/boards/native/nrf_bsim/nrf5340bsim_nrf5340_cpuapp_native.dts - will vanish when old hwmv1 board PR is merged
/tmp/aa/zephyr/boards/native/nrf_bsim/nrf5340bsim_nrf5340_cpunet_native.dts - will vanish when old hwmv1 board PR is merged
/tmp/aa/zephyr/boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuppr_xip.dts
/tmp/aa/zephyr/boards/olimex/olimex_esp32_evb/olimex_esp32_evb_esp32_appcpu.dts - different CPU cluster so no common file
/tmp/aa/zephyr/boards/qemu/arc/qemu_arc_qemu_arc_hs6x.dts - different CPU cluster so no common file
/tmp/aa/zephyr/boards/qemu/arc/qemu_arc_qemu_arc_hs_xip.dts - different CPU cluster so no common file
/tmp/aa/zephyr/boards/qemu/x86/qemu_x86_atom_nopae.dts - was kept as separate file as part of hwmv2 (vendor was told about features and themselves decided not to use it)
/tmp/aa/zephyr/boards/qemu/x86/qemu_x86_atom_xip.dts - was kept as separate file as part of hwmv2 (vendor was told about features and themselves decided not to use it)
```
So again, I find no valid reason for keeping a disjointed feature for a total of 2 board targets that masks real issues when someone can easily just do:
```
#include "base_file.dts"
```

nordicjm commented on cmake/modules/dts.cmake:
```diff
@@ -235,7 +235,7 @@ zephyr_dt_preprocess(
   DEPS_FILE ${DTS_DEPS}
   EXTRA_CPPFLAGS ${DTS_EXTRA_CPPFLAGS}
-  INCLUDE_DIRECTORIES ${DTS_ROOT_SYSTEM_INCLUDE_DIRS}
+  INCLUDE_DIRECTORIES ${DTS_ROOT_SYSTEM_INCLUDE_DIRS};${BOARD_DIR}
```
Reverted

nordicjm commented on cmake/modules/boards.cmake:
```diff
@@ -57,7 +57,7 @@ zephyr_check_cache(BOARD REQUIRED)
 # 'BOARD_ROOT' is a prioritized list of directories where boards may
 # be found. It always includes ${ZEPHYR_BASE} at the lowest priority (except for unittesting).
-if(NOT unittest IN_LIST Zephyr_FIND_COMPONENTS)
+if(NOT unittest IN_LIST Zephyr_FIND_COMPONENTS AND NOT "${ZEPHYR_BASE}" IN_LIST BOARD_ROOT)
```
Reverted

nordicjm committed:
Revert "sample: basic: blinky_pwm: Exclude rpi_pico w variant"

This reverts commit c47c37d3db876f168f33bb3fa8c28ef8976093b6.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on scripts/pylib/pytest-twister-harness/README.rst:
```diff
@@ -21,7 +21,7 @@ Run exemplary test shell application by Twister:
 
   # native_sim & QEMU
-  ./scripts/twister -p native_sim -p qemu_x86 -T samples/subsys/testsuite/pytest/shell
+  ./scripts/twister -p native_sim/native -p qemu_x86 -T samples/subsys/testsuite/pytest/shell
```
Fixed

nordicjm commented on samples/basic/blinky_pwm/sample.yaml:
```diff
@@ -8,3 +8,4 @@ tests:
       - pwm
     depends_on: pwm
     harness: led
+    platform_exclude: rpi_pico/rp2040/w
```
Fixed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -1593,20 +1593,17 @@ function(zephyr_build_string outvar)
-    endwhile()
+
+    list(LENGTH str_segment_list str_segment_list_len)
+    if("${str_segment_list_len}" EQUAL "2")
```
Fixed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -1593,20 +1593,17 @@ function(zephyr_build_string outvar)
   string(JOIN "_" ${outvar} ${str_segment_list} ${revision_string} ${BUILD_STR_BUILD})
 
   if(BUILD_STR_MERGE)
```
Fixed

nordicjm committed:
doc: hardware: porting: board_porting: Remove common include text

Removes information on common includes as this feature has been
removed

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
Running tests locally to work around the github workflow restriction - compliance:
```
git config diff.renameLimit 10000
./scripts/ci/check_compliance.py --annotate -e KconfigBasic -c 0bcf71e28a28b62d186a9f9495a56adb1e4aed9f..
Running Pylint           tests in /tmp/aa/zephyr ...
Running KconfigHWMv2     tests in /tmp/aa/zephyr ...
Running KeepSorted       tests in /tmp/aa/zephyr ...
Running Gitlint          tests in /tmp/aa/zephyr ...
Skipping KconfigBasic
Running ModulesMaintainers tests in /tmp/aa/zephyr ...
Running Nits             tests in /tmp/aa/zephyr ...
Running ImageSize        tests in /tmp/aa/zephyr ...
Running MaintainersFormat tests in /tmp/aa/zephyr ...
Running KconfigBasicNoModules tests in /tmp/aa/zephyr ...
warning: LV_USE_GPU_STM32_DMA2D (defined at boards/st/stm32h7b3i_dk/Kconfig.defconfig:25) defined without a type
warning: LV_GPU_DMA2D_CMSIS_INCLUDE (defined at boards/st/stm32h7b3i_dk/Kconfig.defconfig:28) defined without a type
Running Identity         tests in /tmp/aa/zephyr ...
Running Kconfig          tests in /tmp/aa/zephyr ...
Running YAMLLint         tests in /tmp/aa/zephyr ...
Running GitDiffCheck     tests in /tmp/aa/zephyr ...
Running DevicetreeBindings tests in /tmp/aa/zephyr ...
Running BinaryFiles      tests in /tmp/aa/zephyr ...
Running BoardYml         tests in /tmp/aa/zephyr ...
Running Checkpatch       tests in /tmp/aa/zephyr ...

Complete results in compliance.xml

git log --oneline
ef997f4d0e8 (HEAD -> hwmv2-revert-common-board, mine/hwmv2-revert-common-board) doc: hardware: porting: board_porting: Remove common include text
```

nordicjm commented:
Running tests on brancj to work around the github workflow restriction - clang https://github.com/thedjnK/zephyr/actions/runs/8456609652/job/23166925452?pr=1:
```
INFO    - /__w/zephyr/zephyr/twister-out/native_sim_native/samples/subsys/tracing/sample.tracing.user/handler.log
INFO    - 2744 test scenarios (895 test instances) selected, 144 configurations skipped (0 by static filter, 144 at runtime).
INFO    - 748 of 895 test configurations passed (99.60%), 1 failed, 2 errored, 144 skipped with 0 warnings in 1585.82 seconds
INFO    - In total 6503 test cases were executed, 1364 skipped on 0 out of total 706 platforms (0.00%)
INFO    - 608 test configurations executed on platforms, 143 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /__w/zephyr/zephyr/twister-out/twister.json
INFO    - Writing xunit report /__w/zephyr/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /__w/zephyr/zephyr/twister-out/twister_report.xml...
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) samples/subsys/display/lvgl/sample.display.lvgl.gui on native_sim/native error (Cmake build failure)
INFO    - 2) samples/subsys/tracing/sample.tracing.user on native_sim/native failed (Timeout)
INFO    - 3) samples/modules/lvgl/accelerometer_chart/sample.modules.lvgl.accelerometer_chart on native_sim/native error (Cmake build failure)
INFO    - 
INFO    - To rerun the tests, call twister using the following commandline:
INFO    - west twister -p <PLATFORM> -s <TEST ID>, for example:
INFO    - 
INFO    - west twister -p native_sim/native -s samples/modules/lvgl/accelerometer_chart/sample.modules.lvgl.accelerometer_chart
INFO    - or with west:
INFO    - west build -p -b native_sim/native samples/modules/lvgl/accelerometer_chart -T sample.modules.lvgl.accelerometer_chart
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
INFO    - Run completed
```
These failed due to missing renames in overlays which have been updated here, running those locally:
```
ZEPHYR_TOOLCHAIN_VARIANT=llvm ./scripts/twister -T samples/subsys/display/lvgl/ -p native_sim/native
ZEPHYR_BASE unset, using "/tmp/aa/zephyr"
Renaming output directory to /tmp/aa/zephyr/twister-out.5
INFO    - Using Ninja..
INFO    - Zephyr version: v3.6.0-1850-gd5d59603e64a
INFO    - Using 'llvm' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /tmp/aa/zephyr/twister-out/testplan.json
INFO    - JOBS: 8
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    3/   3  100%  skipped:    2, failed:    0, error:    0
INFO    - 3 test scenarios (3 test instances) selected, 2 configurations skipped (2 by static filter, 0 at runtime).
INFO    - 1 of 3 test configurations passed (100.00%), 0 failed, 0 errored, 2 skipped with 0 warnings in 29.60 seconds
INFO    - In total 0 test cases were executed, 3 skipped on 1 out of total 706 platforms (0.14%)
INFO    - 0 test configurations executed on platforms, 1 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /tmp/aa/zephyr/twister-out/twister.json
INFO    - Writing xunit report /tmp/aa/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /tmp/aa/zephyr/twister-out/twister_report.xml...
INFO    - Run completed

ZEPHYR_TOOLCHAIN_VARIANT=llvm ./scripts/twister -T samples/subsys/tracing/ -p native_sim/native
ZEPHYR_BASE unset, using "/tmp/aa/zephyr"
Renaming output directory to /tmp/aa/zephyr/twister-out.6
INFO    - Using Ninja..
INFO    - Zephyr version: v3.6.0-1850-gd5d59603e64a
INFO    - Using 'llvm' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /tmp/aa/zephyr/twister-out/testplan.json
INFO    - JOBS: 8
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:   10/  10  100%  skipped:    7, failed:    0, error:    0
INFO    - 10 test scenarios (10 test instances) selected, 7 configurations skipped (7 by static filter, 0 at runtime).
INFO    - 3 of 10 test configurations passed (100.00%), 0 failed, 0 errored, 7 skipped with 0 warnings in 24.64 seconds
INFO    - In total 3 test cases were executed, 7 skipped on 1 out of total 706 platforms (0.14%)
INFO    - 3 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /tmp/aa/zephyr/twister-out/twister.json
INFO    - Writing xunit report /tmp/aa/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /tmp/aa/zephyr/twister-out/twister_report.xml...
INFO    - Run completed

ZEPHYR_TOOLCHAIN_VARIANT=llvm ./scripts/twister -T samples/modules/lvgl/accelerometer_chart -p native_sim/native
ZEPHYR_BASE unset, using "/tmp/aa/zephyr"
Renaming output directory to /tmp/aa/zephyr/twister-out.8
INFO    - Using Ninja..
INFO    - Zephyr version: v3.6.0-1850-gd5d59603e64a
INFO    - Using 'llvm' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /tmp/aa/zephyr/twister-out/testplan.json
INFO    - JOBS: 8
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0, error:    0
INFO    - 1 test scenarios (1 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 errored, 0 skipped with 0 warnings in 28.07 seconds
INFO    - In total 0 test cases were executed, 1 skipped on 1 out of total 706 platforms (0.14%)
INFO    - 0 test configurations executed on platforms, 1 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /tmp/aa/zephyr/twister-out/twister.json
INFO    - Writing xunit report /tmp/aa/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /tmp/aa/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

This is a merge heck PR and constantly needs rebases, can we get this approved and merged so this doesn't keep need to happening?

nashif reviewed (changes requested):
This is a massive change after migration to V2 just less than 1 month ago and it will cause lots of breakges and a change of board names and how they are referennced. Needs tmore discussion.

nordicjm commented:
> This is a massive change after migration to V2 just less than 1 month ago and it will cause lots of breakges and a change of board names and how they are referennced. Needs tmore discussion.

Discussion with who/where? I pinged you twice on discord on the 21st March about this PR and also asked others to take a look, I then brought it up at the architecture working group on 26th March and discussed it there **twice** to ensure everyone understood, a meeting which you yourself were present at, there were no objections raised. This follows the zephyr process so I'm a bit unsure what else is expected

nordicjm committed:
samples/tests: Rename boards

Renames boards to use identifiers which have the SoC name in, also
fixes some test issues that were noticed

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
docs: Update include files

Updates rst files to include renamed files with SoC names in them

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: Rename files to have SoCs in name

Adds the SoC name to files

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
samples/tests: Rename files to have SoCs in name

Adds the SoC name to files

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
snippets: Rename files to have SoCs in name

Adds the SoC name to files

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
testsuite: Add SoC to unit_testing board

Adds the SoC name to the unit testing board

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

erwango commented:
I understand this change requires more discussion, but how can we proceed for PR which are currently blocked by CI issues related to this ? (see https://github.com/zephyrproject-rtos/zephyr/pull/70434 for instance)

tejlmand reviewed (changes requested):
Code doesn't pick up deprecated filenames as expected.
See specific comment.

tejlmand commented on cmake/modules/dts.cmake:
```diff
@@ -123,13 +123,21 @@ set(DTS_CMAKE                   ${PROJECT_BINARY_DIR}/dts.cmake)
   foreach(str ${dts_board_string})
     if(EXISTS ${BOARD_DIR}/${str}.dts)
+      if("${str}" IN_LIST deprecated_file_names)
+        message(DEPRECATION
+                "Devicetree files (${str}.dts) without the SoC name are deprecated"
+                " after Zephyr 3.6, you should use <board>_<soc>.dts instead")
```
This warning is also printed for boards described in HWMv1 format, but a board in v1 format must use this filename format, which is `<board>.dts`, and therefore this warning shouldn't be printed for hwmv1.

tejlmand replied:
Fixed

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1605,6 +1613,10 @@ function(zephyr_build_string outvar)
     list(APPEND ${outvar} "${BUILD_STR_BOARD}${qualifiers_without_soc}")
+
+    if(BUILD_STR_DEPRECATED)
+      set(${BUILD_STR_DEPRECATED} ${BUILD_STR_BOARD} PARENT_SCOPE)
```
Fixed

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1593,20 +1593,18 @@ function(zephyr_build_string outvar)
+    list(LENGTH str_segment_list str_segment_list_len)
+
+    # Include board name without SoC as deprecated option
+    string(REGEX REPLACE "^/(.+)/" "/" qualifiers_without_soc "${BUILD_STR_BOARD_QUALIFIERS}")
```
Fixed

evgeniy-paltsev reviewed (changes requested):
Is there a chance to avoid such board renames?

```diff
-rename from boards/snps/hsdk/hsdk_defconfig
+rename to boards/snps/hsdk/hsdk_arc_hsdk_defconfig

-rename from boards/snps/hsdk4xd/hsdk4xd_defconfig
+rename to boards/snps/hsdk4xd/hsdk4xd_arc_hsdk4xd_defconfig
```

New names look really weird:
`hsdk_arc_hsdk`
`hsdk4xd_arc_hsdk4xd`

carlescufi commented:
> Is there a chance to avoid such board renames?

@evgeniy-paltsev there is, but we decided not to go with it for the time being to avoid adding more complexity to this already complex PR. See the last slide in my presentation here:
https://docs.google.com/presentation/d/1BCeKKhs3BV5EciA5cQrbEfMexOjWsHyj-dgcB0yJrbQ/edit#slide=id.g2c87032e9ae_0_39

This was discussed in the Architecture WG and in a separate Discord thread: https://discord.com/channels/720317445772017664/1224744062687641622/1224744064168235028

tejlmand commented:
> Is there a chance to avoid such board renames?

There is, and it's part of this PR.
But at the latest Arch WG meeting it was decided to go for a full rename in Zephyr tree, and deprecate the short form.

If you don't rename `hsdk4xd_defconfig`, then this is what you'll see during build:
```
$ cmake -GNinja -DBOARD=hsdk4xd/arc_hsdk4xd -Bbuild samples/hello_world/
Loading Zephyr default modules (Zephyr repository).
-- Application: /projects/github/ncs/zephyr/samples/hello_world
...
-- Board: hsdk4xd, qualifiers: arc_hsdk4xd
...
CMake Deprecation Warning at /projects/github/ncs/zephyr/cmake/modules/FindDeprecated.cmake:240 (message):
  defconfig files (hsdk4xd_defconfig) without the SoC name are deprecated
  after Zephyr 3.6, you should use <board>_<soc>_defconfig instead
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/modules/extensions.cmake:2709 (find_package)
  /projects/github/ncs/zephyr/cmake/modules/kconfig.cmake:78 (zephyr_file)
  /projects/github/ncs/nrf/cmake/modules/kconfig.cmake:29 (include)
  /projects/github/ncs/zephyr/cmake/modules/zephyr_default.cmake:132 (include)
  /projects/github/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /projects/github/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:117 (include_boilerplate)
  CMakeLists.txt:5 (find_package)


Parsing /projects/github/ncs/zephyr/samples/hello_world/Kconfig
Loaded configuration '/projects/github/ncs/zephyr/boards/snps/hsdk4xd/hsdk4xd_defconfig'
Merged configuration '/projects/github/ncs/zephyr/samples/hello_world/prj.conf'
Configuration saved to '/projects/github/ncs/zephyr/build/zephyr/.config'
...
```

So we could simply remove the deprecation part (and the renames).
For the CMake part, I would move some code around, because the current deprecated code is now moved to the deprecated modules.

evgeniy-paltsev reviewed (approved):
Hi,

I've checked the changes in detail - looks like I was confused and we are only changing board files names, but the board itself still would be accessible with the previous name.
In that case I don't have anything against the change.

Sorry for the noise.
